File: Problems/Leetcode/1001-2000/1301-1400/1319.py
from typing import List


class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        pa = list(range(n))
        size = [1] * n
        cnt = 0

        def find(x: int) -> int:
            if pa[x] != x:
                pa[x] = find(pa[x])
            return pa[x]

        def merge(x: int, y: int) -> None:
            px, py = find(x), find(y)
            if px == py:
                nonlocal cnt
                cnt += 1  # 记录多余的线
                return
            if size[px] > size[py]:
                px, py = py, px
            pa[px] = py
            size[py] += size[px]

        for con in connections:
            merge(con[0], con[1])

        s = set(find(i) for i in range(n))
        return len(s) - 1 if cnt >= len(s) - 1 else -1  # 操作次数等于连通分量个数 -1

File: Problems/Leetcode/1001-2000/1301-1400/1319.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int makeConnected(int n, vector<vector<int>>& connections) {
        vector<int> pa(n), size(n, 1);
        iota(pa.begin(), pa.end(), 0);
        int cnt = 0; // 记录多余的线

        function<int(int)> find = [&](int x) -> int {
            return pa[x] == x ? x : pa[x] = find(pa[x]);
        };

        for (auto& vec : connections) {
            int px = find(vec[0]), py = find(vec[1]);
            if (px == py) {
                cnt++;
                continue;
            }
            if (size[px] > size[py]) {
                swap(px, py);
            }
            pa[px] = py;
            size[py] += size[px];
        }

        set<int> set;
        for (int i = 0; i < n; ++i) {
            set.emplace(find(i));
        }
        return cnt >= set.size() - 1 ? set.size() - 1 : -1;
    }
};

int main() {
    Solution sol;

    return 0;
}